juju/juju
# Add params to application service create app api and rename generic Save methods.

The initial bare bones application service needs to be extended to take additional params when creating an application - these are currently empty but will be used later to hold constraints, storage etc.

Making the above change, the generic "Save" method was replaced with properly named CreateApplication and UpsertCAASUnit methods to reflect the interface we want to expose. The underlying skeleton state calls don't need to be changed yet. The "Save" change was also done on the machine service; it's now "CreateMachine". 

The applicsation facade had methods which took a large number of parameters, but these values were available on the struct. So this has been cleaned up so we don't keep increasing the param count as we need to pass in more services.

The state AddApplication method was taking an application service as a parameter to allow for dual writing to mongo and dqlite. Turns out this will make it harder to deal with storage, so instead the facade does the dual writing. To allow this, state gains a new public method "ReadSequence" to allow the facade to read the next unit number from mongo - this is temporary and will disappear once more is moved across to dqlite.

## QA steps

bootstrap 
enable-ha
deploy ubuntu -n 2
add-unit ubuntu -n 2

open a dqlite shell and select from the machine, unit and application tables and check contents matches mongo.

Repeat on k8s with snappass-test charm.

## Links

**Jira card:** JUJU-5599
